library/core/src/any.rs

```diff
@@ -1034,6 +1034,89 @@ impl<'a> Demand<'a> {
     pub fn is_satisfied(&self) -> bool {
         self.0.is_satisfied()
     }
+
+    /// Check if the `Demand` would be satisfied if provided with the
+    /// specified type. If the type does not match or has already been
+    /// provided, returns false.
+    ///
+    /// # Examples
+    ///
+    /// Check if an `u8` still needs to be provided and then provides
+    /// it.
+    ///
+    /// ```rust
+    /// # #![feature(provide_any)]
+    /// use std::any::{Provider, Demand};
+    /// # struct SomeConcreteType;
+    /// # fn expensive_computation() -> Option<u8> { Some(42) }
+    ///
+    /// impl Provider for SomeConcreteType {
+    ///     fn provide<'a>(&'a self, demand: &mut Demand<'a>) {
+    ///         if demand.would_be_satisfied_by_value_of::<u8>() {
+    ///             if let Some(value) = expensive_computation() {
+    ///                 demand.provide_value(value);
+    ///             }
+    ///         }
+    ///         if demand.would_be_satisfied_by_value_of::<u8>() {
+    ///             unreachable!("An u8 has already been provided");
+    ///         }
```

Comment:
why is this unreachable? I know `expensive_computation` always returns `Some` but in this example that is hidden, so it seems like this should sometimes be reachable

Reply:
This is something I struggled with for these examples. I need to put _some_ kind of assertion inside of the implementation, otherwise the tests have no possible failure case and lose half of their value.